File: examples/triangles.rs
use anyhow::{Context, Result};
use raydeon::shapes::Triangle;
use raydeon::{Camera, Scene, WPoint3, WVec3};

fn main() -> Result<()> {
    env_logger::Builder::from_default_env()
        .format_timestamp_nanos()
        .init();

    let scene = Scene::new(vec![
        Box::new(Triangle::new(
            WPoint3::new(0.0, 0.0, 0.0),
            WPoint3::new(0.0, 0.0, 1.0),
            WPoint3::new(1.0, 0.0, 1.0),
        )),
        Box::new(Triangle::new(
            WPoint3::new(0.25, 0.25, 0.0),
            WPoint3::new(0.0, 0.25, 1.0),
            WPoint3::new(-0.65, 0.25, 1.0),
        )),
    ]);

    let eye = WPoint3::new(0.0, 3.0, 0.0);
    let focus = WVec3::new(0.0, 0.0, 0.0);

    let look = eye.to_vector().normalize();
    let up = look.cross(WVec3::new(0.0, 0.0, 1.0)).cross(look);

    let fovy = 50.0;
    let width = 1024.0;
    let height = 1024.0;
    let znear = 0.1;
    let zfar = 10.0;

    let camera = Camera::look_at(eye, focus, up).perspective(fovy, width, height, znear, zfar);

    let paths = scene.attach_camera(camera).render();

    // We currently don't have any functionality to aid in emitting SVG images, so you will
    // be required to use the [svg crate.](https://crates.io/crates/svg)
    let mut svg_doc = svg::Document::new()
        .set("width", "8in")
        .set("height", "8in")
        .set("viewBox", (0, 0, width, height))
        .set("stroke-width", "0.7mm")
        .set("stroke", "black")
        .set("fill", "none")
        .add(
            svg::node::element::Rectangle::new()
                .set("x", 0)
                .set("y", 0)
                .set("width", "100%")
                .set("height", "100%")
                .set("fill", "white"),
        );

    // We have to flip the y-axis in our svg...
    let mut item_group = svg::node::element::Group::new()
        .set("transform", format!("translate(0, {}) scale(1,-1)", height));

    for path in paths {
        let (p1, p2) = (path.p1, path.p2);
        item_group = item_group.add(
            svg::node::element::Line::new()
                .set("x1", p1.x)
                .set("y1", p1.y)
                .set("x2", p2.x)
                .set("y2", p2.y),
        );
    }

    svg_doc = svg_doc.add(item_group);

    svg::save("triangles.svg", &svg_doc).context("Failed to write svg")
}

File: pyraydeon/examples/triangles.py
import svg

from pyraydeon import (
    Camera,
    CollisionGeometry,
    Geometry,
    LineSegment3D,
    Point3,
    Scene,
    Stroke,
    Tri,
    Vec3,
)


class CustomTriangle(Geometry):
    def __init__(self, p1: Point3, p2: Point3, p3: Point3) -> None:
        self.tri = Tri(p1, p2, p3)

    def collision_geometry(self) -> list[CollisionGeometry]:
        return self.tri.collision_geometry()

    def paths(self, cam: Camera) -> list[LineSegment3D]:
        return self.tri.paths(cam)


scene = Scene(
    [
        CustomTriangle(
            Point3(0, 0, 0),
            Point3(0, 0, 1),
            Point3(1, 0, 1),
        ),
        Tri(
            Point3(0.25, 0.25, 0),
            Point3(0, 0.25, 1),
            Point3(-0.65, 0.25, 1),
        ),
    ]
)

eye = Point3(0, 3, 0)
focus = Vec3(0, 0, 0)
up = Vec3(0, 0, 1)

fovy = 50.0
width = 1024
height = 1024
znear = 0.1
zfar = 10.0

cam = Camera().look_at(eye, focus, up).perspective(fovy, width, height, znear, zfar)

strokes: list[Stroke] = scene.render(cam)

canvas = svg.SVG(
    width="8in",
    height="8in",
    viewBox="0 0 1024 1024",
)
backing_rect = svg.Rect(
    x=0,
    y=0,
    width="100%",
    height="100%",
    fill="white",
)
svg_lines = [
    svg.Line(
        x1=f"{stroke.p1[0]}",
        y1=f"{stroke.p1[1]}",
        x2=f"{stroke.p2[0]}",
        y2=f"{stroke.p2[1]}",
        stroke_width="0.7mm",
        stroke="black",
    )
    for stroke in strokes
]
line_group = svg.G(transform=f"translate(0, {height}) scale(1, -1)", elements=svg_lines)
canvas.elements = [backing_rect, line_group]


print(canvas)
